dbcon/joblist/tupleconstantonlystep.cpp

```diff
@@ -0,0 +1,472 @@
+ the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+// #define NDEBUG
+#include <cassert>
+#include <sstream>
+#include <iomanip>
+using namespace std;
+
+#include <boost/shared_ptr.hpp>
+
+#include <boost/uuid/uuid_io.hpp>
+using namespace boost;
+
+#include "messagequeue.h"
+using namespace messageqcpp;
+
+#include "loggingid.h"
+#include "errorcodes.h"
+#include "idberrorinfo.h"
+#include "errorids.h"
+using namespace logging;
+
+#include "calpontsystemcatalog.h"
+#include "constantcolumn.h"
+using namespace execplan;
+
+#include "jobstep.h"
+#include "rowgroup.h"
+using namespace rowgroup;
+
+#include "querytele.h"
+using namespace querytele;
+
+#include "jlf_common.h"
+#include "tupleconstantonlystep.h"
+
+namespace joblist
+{
+SJSTEP addConstantStep(const JobInfo& jobInfo, const rowgroup::RowGroup* rg)
+{
+  TupleConstantOnlyStep* tcos = new TupleConstantOnlyStep(jobInfo);
+  tcos->initialize(jobInfo, rg);
+  SJSTEP spcs(tcos);
+  return spcs;
+}
+
+// class TupleConstantOnlyStep
+TupleConstantOnlyStep::TupleConstantOnlyStep(const JobInfo& jobInfo) : JobStep(jobInfo)
+{
+  fExtendedInfo = "TCOS: ";
+  fQtc.stepParms().stepType = StepTeleStats::T_TCOS;
+}
+
+TupleConstantOnlyStep::~TupleConstantOnlyStep()
+{
+}
+
+const RowGroup& TupleConstantOnlyStep::getOutputRowGroup() const
+{
+  return rowGroupOut_;
+}
+
+void TupleConstantOnlyStep::setOutputRowGroup(const rowgroup::RowGroup& rg)
+{
+  throw runtime_error("Disabled, use initialize() to set output RowGroup.");
+}
+
+const RowGroup& TupleConstantOnlyStep::getDeliveredRowGroup() const
+{
+  return rowGroupOut_;
+}
+
+void TupleConstantOnlyStep::deliverStringTableRowGroup(bool b)
+{
+  rowGroupOut_.setUseStringTable(b);
+}
+
+bool TupleConstantOnlyStep::deliverStringTableRowGroup() const
+{
+  return rowGroupOut_.usesStringTable();
+}
+
+void TupleConstantOnlyStep::join()
+{
+}
+
+// void TupleConstantOnlyStep::initialize(const RowGroup& rgIn, const JobInfo& jobInfo)
+void TupleConstantOnlyStep::initialize(const JobInfo& jobInfo, const rowgroup::RowGroup* rgIn)
+{
+  vector<uint32_t> oids;
+  vector<uint32_t> keys;
+  vector<uint32_t> scale;
+  vector<uint32_t> precision;
+  vector<CalpontSystemCatalog::ColDataType> types;
+  vector<uint32_t> csNums;
+  vector<uint32_t> pos;
+  pos.push_back(2);
+
+  deliverStringTableRowGroup(false);
+
+  for (uint64_t i = 0; i < jobInfo.deliveredCols.size(); i++)
+  {
+    const ConstantColumn* cc = dynamic_cast<const ConstantColumn*>(jobInfo.deliveredCols[i].get());
+
+    if (cc == NULL)
+      throw runtime_error("none constant column found.");
+
+    CalpontSystemCatalog::ColType ct = cc->resultType();
+
+    if (ct.colDataType == CalpontSystemCatalog::VARCHAR)
+      ct.colWidth++;
+
+    // Round colWidth up
+    if (ct.colWidth == 3)
+      ct.colWidth = 4;
+    else if (ct.colWidth == 5 || ct.colWidth == 6 || ct.colWidth == 7)
+      ct.colWidth = 8;
+
+    oids.push_back(-1);
```

Comment:
You can construct these two vectors right away with a vec ctor with a default value.